x-pack/plugins/infra/public/components/asset_details/hooks/use_page_header.tsx

```diff
@@ -4,26 +4,88 @@
  * 2.0; you may not use this file except in compliance with the Elastic License
  * 2.0.
  */
-
-import { useEuiTheme, EuiIcon, type EuiPageHeaderProps } from '@elastic/eui';
+import {
+  useEuiTheme,
+  EuiIcon,
+  type EuiPageHeaderProps,
+  type EuiBreadcrumbsProps,
+  EuiFlexGroup,
+  EuiFlexItem,
+} from '@elastic/eui';
 import { css } from '@emotion/react';
 import { useLinkProps } from '@kbn/observability-shared-plugin/public';
 import React, { useCallback, useMemo } from 'react';
 import { capitalize } from 'lodash';
+import { useHistory, useLocation } from 'react-router-dom';
+import { FormattedMessage } from '@kbn/i18n-react';
+import { useKibanaContextForPlugin } from '../../../hooks/use_kibana';
 import { APM_HOST_FILTER_FIELD } from '../constants';
 import { LinkToAlertsRule, LinkToApmServices, LinkToNodeDetails } from '../links';
-import { FlyoutTabIds, type LinkOptions, type Tab, type TabIds } from '../types';
+import { FlyoutTabIds, type RouteState, type LinkOptions, type Tab, type TabIds } from '../types';
 import { useAssetDetailsStateContext } from './use_asset_details_state';
 import { useDateRangeProviderContext } from './use_date_range';
 import { useTabSwitcherContext } from './use_tab_switcher';
 
 type TabItem = NonNullable<Pick<EuiPageHeaderProps, 'tabs'>['tabs']>[number];
 
-export const usePageHeader = (tabs: Tab[], links?: LinkOptions[]) => {
+export const usePageHeader = (tabs: Tab[] = [], links: LinkOptions[] = []) => {
   const { rightSideItems } = useRightSideItems(links);
   const { tabEntries } = useTabs(tabs);
+  const { breadcrumbs } = useTemplateHeaderBreadcrumbs();
+
+  return { rightSideItems, tabEntries, breadcrumbs };
+};
+
+export const useTemplateHeaderBreadcrumbs = () => {
+  const history = useHistory();
+  const location = useLocation<RouteState>();
+  const {
+    services: {
+      application: { navigateToApp },
+    },
+  } = useKibanaContextForPlugin();
+
+  const onClick = (e: React.MouseEvent) => {
+    if (location.state) {
+      navigateToApp(location.state.originAppId, {
+        replace: true,
+        path: `${location.state.originPathname}${location.state.originData}`,
+      });
+    } else {
+      history.goBack();
+    }
+    e.preventDefault();
+  };
+
+  const breadcrumbs: EuiBreadcrumbsProps['breadcrumbs'] =
+    // If there is a state object in location, it's persisted in case the page is opened in a new tab or after page refresh
+    // With that, we can show the return button. Otherwise, it will be hidden (ex: the user opened a shared URL or opened the page from their bookmarks)
+    location.state || history.length > 1
```

Comment:
One small caveat about `history.length` here. If a user has a browser tab with existing history and then pastes a Kibana URL into that tab, `history` will still have all of the previous records and the Return button will go back to an arbitrary page from the history. Kind of an edge case, but worth mentioning.

Reply:
I tried to simulate this, but the `history` is reset when I paste a URL in a tab with existing `history`.
Returning using `history.goBack()` should be an edge case. I'm even thinking about removing it.